cpp/src/reshape/explode.cu

```diff
@@ -0,0 +1,140 @@
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/exec_policy.hpp>
+
+#include <thrust/binary_search.h>
+#include <thrust/iterator/counting_iterator.h>
+#include <thrust/iterator/transform_iterator.h>
+
+#include <memory>
+#include <type_traits>
+
+namespace cudf {
+namespace detail {
+namespace {
+/**
+ * @brief Function object for exploding a column.
+ */
+struct explode_functor {
+  template <typename T>
+  std::unique_ptr<table> operator()(table_view const& input_table,
+                                    size_type explode_column_idx,
+                                    rmm::cuda_stream_view stream,
+                                    rmm::mr::device_memory_resource* mr) const
+  {
+    CUDF_FAIL("Unsupported non-list column");
+
+    return std::make_unique<table>();
+  }
+};
+
+template <>
+std::unique_ptr<table> explode_functor::operator()<list_view>(
+  table_view const& input_table,
+  size_type explode_column_idx,
+  rmm::cuda_stream_view stream,
+  rmm::mr::device_memory_resource* mr) const
+{
+  /* we explode by building a gather map that includes the number of entries in each list inside
+   the column for each index. Interestingly, this can be done with lower_bound across the offsets
+   as values between the offsets will all map down to the index below. We have some off-by-one
+   manipulations we need to do with the output, but it's almost our gather map by itself. Once we
+   build the gather map we need to remove the explode column from the table and run gather on it.
+   Next we build the explode column, which turns out is simply lifting the child column out of the
+   explode column. This unrolls the top level of lists. Then we need to insert the explode column
+   back into the table and return it. */
+  lists_column_view lc{input_table.column(explode_column_idx)};
+  rmm::device_uvector<size_type> gather_map_indices(lc.child().size(), stream, mr);
+  auto offsets = lc.offsets();
+
+  auto offsets_minus_one = thrust::make_transform_iterator(offsets.begin<size_type>(),
+                                                           [] __device__(auto i) { return i - 1; });
+  auto counting_iter     = thrust::make_counting_iterator(0);
+
+  thrust::lower_bound(rmm::exec_policy(stream),
+                      offsets_minus_one,
+                      offsets_minus_one + offsets.size(),
+                      counting_iter,
+                      counting_iter + gather_map_indices.size(),
+                      gather_map_indices.begin());
+
+  auto select_iter = thrust::make_transform_iterator(
+    thrust::make_counting_iterator(0),
+    [explode_column_idx](int i) { return i >= explode_column_idx ? i + 1 : i; });
+  std::vector<int> selected_columns(select_iter, select_iter + input_table.num_columns() - 1);
+
+  auto gather_map_iter = thrust::make_transform_iterator(gather_map_indices.begin(),
+                                                         [] __device__(int i) { return i - 1; });
```

Comment:
This `i - 1` could be avoided if `offsets_minus_one + 1` is passed as first to `thrust::lower_bound`.
(iterator version of `cudf::detail::gather` could be replaced with column_view version. It might save some compile time).

Reply:
This is a great piece of advice. Thank you for taking the time to really get in here and understand the intent. I was trying to figure out ways to remove some of the -1 operations here and couldn't see that.